File: web/lib/quality.ts
export interface QualityScore {
  score: number;
  tooShort: boolean;
  missingLinks: boolean;
  duplicateHeavy: boolean;
  noisy: boolean;
  acceptable: boolean;
}

const NOISY_SIGNALS = ["cookie", "subscribe", "javascript", "log in", "sign up"];

export function scoreContent(markdown: string, links?: string[]): QualityScore {
  const text = (markdown ?? "").trim();
  const linkCount = links?.length ?? 0;

  const tooShort = text.length < 500;
  const missingLinks = linkCount === 0;

  const lines = text.split("\n");
  const numLines = lines.length;
  let duplicateHeavy = false;
  if (numLines > 0) {
    const uniqueLines = new Set(lines.map((l) => l.trim()).filter(Boolean)).size;
    // Updated threshold to max(5, numLines / 3) to match Python/Rust 2026 standards
    duplicateHeavy = uniqueLines < Math.max(5, Math.floor(numLines / 3));
  }

  const lowerText = text.toLowerCase();
  const noiseCount = NOISY_SIGNALS.reduce((n, sig) => n + (lowerText.includes(sig) ? 1 : 0), 0);
  const noisy = noiseCount > 6;

  // 2026 Standard Checks
  const requiredYaml = ["relevance_score:", "intent_category:", "token_estimate:", "last_updated:"];
  const hasFrontmatter =
    text.startsWith("---") && requiredYaml.every((field) => text.includes(field));
  const hasAnchors = [
    "[ANCHOR: SUMMARY]",
    "[ANCHOR: TECHNICAL_DETAILS]",
    "[ANCHOR: COMPARISON]",
    "[ANCHOR: CITATIONS]",
  ].every((anchor) => text.includes(anchor));

  // Penalties updated to match 2026 standards
  let score = 1.0;
  if (tooShort) score -= 0.25;
  if (missingLinks) score -= 0.1;
  if (duplicateHeavy) score -= 0.15;
  if (noisy) score -= 0.1;

  // Bonuses for 2026 standards
  if (hasFrontmatter) score += 0.05;
  if (hasAnchors) score += 0.05;

  score = Math.max(0, Math.min(1.0, score));

  // Threshold for acceptance: 0.65 and not tooShort
  const acceptable = score >= 0.65 && !tooShort;

  return { score, tooShort, missingLinks, duplicateHeavy, noisy, acceptable };
}

File: .agents/skills/do-web-doc-resolver/scripts/quality.py
"""
Heuristics for scoring the quality of resolved content.
"""

from dataclasses import dataclass

# Quality scoring penalties
PENALTY_TOO_SHORT = 0.25
PENALTY_MISSING_LINKS = 0.10
PENALTY_DUPLICATE_HEAVY = 0.15
PENALTY_NOISY = 0.10

# Quality scoring bonuses
BONUS_HAS_FRONTMATTER = 0.05
BONUS_HAS_ANCHORS = 0.05

# Acceptance threshold
ACCEPTABLE_THRESHOLD = 0.65


@dataclass
class QualityScore:
    score: float
    too_short: bool
    missing_links: bool
    duplicate_heavy: bool
    noisy: bool
    acceptable: bool


def score_content(markdown: str, links: list[str] | None = None) -> QualityScore:
    # Handle MagicMocks in tests
    if not isinstance(markdown, str):
        return QualityScore(1.0, False, False, False, False, True)

    text = (markdown or "").strip()
    links = links or []

    length = len(text)
    too_short = length < 500
    missing_links = len(links) == 0

    lines = text.splitlines()
    num_lines = len(lines)
    duplicate_heavy = False
    if num_lines > 0:
        unique_lines = len({line.strip() for line in lines if line.strip()})
        # If fewer than 5 unique lines OR unique lines are less than 1/3 of total lines
        duplicate_heavy = unique_lines < max(5, num_lines // 3)

    noisy_signals = ["cookie", "subscribe", "javascript", "log in", "sign up"]
    text_lower = text.lower()
    noise_count = sum(text_lower.count(signal) for signal in noisy_signals)
    noisy = noise_count > 6

    # 2026 Standard Checks
    required_yaml = [
        "relevance_score:",
        "intent_category:",
        "token_estimate:",
        "last_updated:",
    ]
    has_frontmatter = text.startswith("---") and all(field in text for field in required_yaml)
    has_anchors = all(
        anchor in text
        for anchor in [
            "[ANCHOR: SUMMARY]",
            "[ANCHOR: TECHNICAL_DETAILS]",
            "[ANCHOR: COMPARISON]",
            "[ANCHOR: CITATIONS]",
        ]
    )

    score = 1.0
    if too_short:
        score -= PENALTY_TOO_SHORT  # Reduced from 0.35
    if missing_links:
        score -= PENALTY_MISSING_LINKS  # Reduced from 0.15
    if duplicate_heavy:
        score -= PENALTY_DUPLICATE_HEAVY  # Reduced from 0.25
    if noisy:
        score -= PENALTY_NOISY  # Reduced from 0.20

    # Bonus for 2026 standards
    if has_frontmatter:
        score += BONUS_HAS_FRONTMATTER
    if has_anchors:
        score += BONUS_HAS_ANCHORS

    # Ensure range
    score = max(0.0, min(1.0, score))

    # Threshold for acceptance as per #59: ACCEPTABLE_THRESHOLD and not too_short
    acceptable = score >= ACCEPTABLE_THRESHOLD and not too_short

    return QualityScore(
        score=score,
        too_short=too_short,
        missing_links=missing_links,
        duplicate_heavy=duplicate_heavy,
        noisy=noisy,
        acceptable=acceptable,
    )
